File: tools/retry_manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
重试管理器 (Retry Manager)
用于限制失败操作的重试次数，避免无意义的重复尝试

使用场景：
- 编译失败
- 测试失败
- 命令执行失败
"""

import os
import sys
import time
import logging
from typing import Callable, Any, Optional
from datetime import datetime


class RetryManager:
    """
    重试管理器

    核心原则：
    - 最多重试 1 次（总共尝试 2 次）
    - 失败时记录详细日志
    - 不要相信"下一次会更好"
    """

    def __init__(self,
                 max_retries: int = 1,
                 log_file: str = "retry_failures.log",
                 enable_logging: bool = True):
        """
        初始化重试管理器

        参数说明：
        - max_retries: 最大重试次数（默认 1，即总共尝试 2 次）
        - log_file: 日志文件路径
        - enable_logging: 是否启用日志记录
        """
        self.max_retries = max_retries
        self.log_file = log_file
        self.enable_logging = enable_logging

        # 设置日志
        if self.enable_logging:
            self._setup_logging()

    def _setup_logging(self):
        """配置日志系统"""
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s [%(levelname)s] %(message)s',
            handlers=[
                logging.FileHandler(self.log_file, encoding='utf-8'),
                logging.StreamHandler(sys.stderr)  # 输出到 stderr，避免破坏 GTP 协议
            ]
        )
        self.logger = logging.getLogger(__name__)

    def execute_with_retry(self,
                           operation: Callable[[], Any],
                           operation_name: str,
                           context: Optional[dict] = None) -> tuple[bool, Any]:
        """
        执行操作，失败时自动重试

        参数说明：
        - operation: 要执行的操作（一个函数）
        - operation_name: 操作名称（用于日志）
        - context: 额外的上下文信息（可选）

        返回值：
        - (成功标志, 结果)
        - 如果失败，返回 (False, 错误信息)
        """
        attempt = 0
        last_error = None

        while attempt <= self.max_retries:
            try:
                # 记录尝试
                if attempt > 0:
                    self._log_retry(operation_name, attempt, context)
                else:
                    self._log_start(operation_name, context)

                # 执行操作
                result = operation()

                # 成功
                self._log_success(operation_name, attempt, context)
                return (True, result)

            except Exception as e:
                last_error = e
                attempt += 1

                # 记录失败
                self._log_failure(operation_name, attempt, e, context)

                # 判断是否继续重试
                if attempt > self.max_retries:
                    break

                # 短暂延迟后重试（避免连续失败）
                time.sleep(0.5)

        # 所有尝试都失败了
        self._log_final_failure(operation_name, self.max_retries + 1, last_error, context)
        return (False, last_error)

    def _log_start(self, operation_name: str, context: Optional[dict]):
        """记录操作开始"""
        msg = f"开始执行: {operation_name}"
        if context:
            msg += f" | 上下文: {context}"
        if self.enable_logging:
            self.logger.info(msg)
        else:
            print(msg, file=sys.stderr)

    def _log_retry(self, operation_name: str, attempt: int, context: Optional[dict]):
        """记录重试"""
        msg = f"重试 [{attempt}/{self.max_retries}]: {operation_name}"
        if context:
            msg += f" | 上下文: {context}"
        if self.enable_logging:
            self.logger.warning(msg)
        else:
            print(f"[警告] {msg}", file=sys.stderr)

    def _log_success(self, operation_name: str, attempt: int, context: Optional[dict]):
        """记录成功"""
        msg = f"成功: {operation_name}"
        if attempt > 0:
            msg += f" (经过 {attempt} 次重试)"
        if context:
            msg += f" | 上下文: {context}"
        if self.enable_logging:
            self.logger.info(msg)
        else:
            print(msg, file=sys.stderr)

    def _log_failure(self, operation_name: str, attempt: int, error: Exception, context: Optional[dict]):
        """记录失败"""
        msg = f"失败 [{attempt}/{self.max_retries + 1}]: {operation_name}"
        msg += f" | 错误: {type(error).__name__}: {str(error)}"
        if context:
            msg += f" | 上下文: {context}"
        if self.enable_logging:
            self.logger.error(msg)
        else:
            print(f"[错误] {msg}", file=sys.stderr)

    def _log_final_failure(self, operation_name: str, total_attempts: int, error: Exception, context: Optional[dict]):
        """记录最终失败"""
        msg = f"==============================\n"
        msg += f"最终失败: {operation_name}\n"
        msg += f"总尝试次数: {total_attempts}\n"
        msg += f"最后错误: {type(error).__name__}: {str(error)}\n"
        if context:
            msg += f"上下文: {context}\n"
        msg += f"时间: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n"
        msg += f"==============================\n"

        if self.enable_logging:
            self.logger.critical(msg)
        else:
            print(f"\n[严重错误]\n{msg}", file=sys.stderr)


def test_retry_manager():
    """测试重试管理器"""
    manager = RetryManager(max_retries=1, log_file="test_retry.log")

    # 测试 1：成功的操作
    def success_operation():
        print("执行成功的操作")
        return "成功结果"

    success, result = manager.execute_with_retry(success_operation, "测试成功操作")
    assert success == True
    assert result == "成功结果"
    print(f"[OK] 测试 1 通过: {result}")

    # 测试 2：第一次失败，第二次成功
    attempt_counter = [0]
    def retry_once_operation():
        attempt_counter[0] += 1
        if attempt_counter[0] == 1:
            raise RuntimeError("第一次尝试失败")
        return "第二次成功"

    success, result = manager.execute_with_retry(retry_once_operation, "测试重试一次")
    assert success == True
    assert result == "第二次成功"
    print(f"[OK] 测试 2 通过: {result}")

    # 测试 3：总是失败
    def always_fail_operation():
        raise RuntimeError("总是失败的操作")

    success, result = manager.execute_with_retry(always_fail_operation, "测试总是失败", {"key": "value"})
    assert success == False
    assert isinstance(result, RuntimeError)
    print(f"[OK] 测试 3 通过: 正确处理了失败")

    print("\n所有测试通过！")


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--test":
        test_retry_manager()
    else:
        print("重试管理器模块")
        print("使用方法: python retry_manager.py --test")

File: lizzieyzy-main/src/main/java/featurecat/lizzie/util/RetryManager.java
package featurecat.lizzie.util;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

/**
 * 重试管理器 (Retry Manager) 用于限制失败操作的重试次数，避免无意义的重复尝试
 *
 * <p>使用场景： - GTP 命令通信失败 - 引擎响应超时 - 文件读写失败
 *
 * <p>核心原则： - 最多重试 1 次（总共尝试 2 次） - 失败时记录详细日志 - 不要相信"下一次会更好"
 */
public class RetryManager {
  private final int maxRetries;
  private final String logFile;
  private final boolean enableLogging;
  private final DateTimeFormatter dateFormatter;

  /**
   * 构造函数
   *
   * @param maxRetries 最大重试次数（默认 1，即总共尝试 2 次）
   * @param logFile 日志文件路径
   * @param enableLogging 是否启用日志记录
   */
  public RetryManager(int maxRetries, String logFile, boolean enableLogging) {
    this.maxRetries = maxRetries;
    this.logFile = logFile;
    this.enableLogging = enableLogging;
    this.dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
  }

  /** 默认构造函数（最多重试 1 次，启用日志） */
  public RetryManager() {
    this(1, "retry_failures.log", true);
  }

  /** 执行操作的接口 */
  @FunctionalInterface
  public interface Operation<T> {
    T execute() throws Exception;
  }

  /** 执行结果 */
  public static class RetryResult<T> {
    public final boolean success;
    public final T result;
    public final Exception error;

    private RetryResult(boolean success, T result, Exception error) {
      this.success = success;
      this.result = result;
      this.error = error;
    }

    public static <T> RetryResult<T> success(T result) {
      return new RetryResult<>(true, result, null);
    }

    public static <T> RetryResult<T> failure(Exception error) {
      return new RetryResult<>(false, null, error);
    }
  }

  /**
   * 执行操作，失败时自动重试
   *
   * @param operation 要执行的操作
   * @param operationName 操作名称（用于日志）
   * @param context 额外的上下文信息（可选）
   * @return 执行结果
   */
  public <T> RetryResult<T> executeWithRetry(
      Operation<T> operation, String operationName, Map<String, String> context) {

    int attempt = 0;
    Exception lastError = null;

    while (attempt <= maxRetries) {
      try {
        // 记录尝试
        if (attempt > 0) {
          logRetry(operationName, attempt, context);
        } else {
          logStart(operationName, context);
        }

        // 执行操作
        T result = operation.execute();

        // 成功
        logSuccess(operationName, attempt, context);
        return RetryResult.success(result);

      } catch (Exception e) {
        lastError = e;
        attempt++;

        // 记录失败
        logFailure(operationName, attempt, e, context);

        // 判断是否继续重试
        if (attempt > maxRetries) {
          break;
        }

        // 短暂延迟后重试（避免连续失败）
        try {
          Thread.sleep(500);
        } catch (InterruptedException ie) {
          Thread.currentThread().interrupt();
          break;
        }
      }
    }

    // 所有尝试都失败了
    logFinalFailure(operationName, maxRetries + 1, lastError, context);
    return RetryResult.failure(lastError);
  }

  /** 简化版本（无上下文） */
  public <T> RetryResult<T> executeWithRetry(Operation<T> operation, String operationName) {
    return executeWithRetry(operation, operationName, null);
  }

  /** 记录操作开始 */
  private void logStart(String operationName, Map<String, String> context) {
    String msg = "开始执行: " + operationName;
    if (context != null && !context.isEmpty()) {
      msg += " | 上下文: " + context;
    }
    log("INFO", msg);
  }

  /** 记录重试 */
  private void logRetry(String operationName, int attempt, Map<String, String> context) {
    String msg = String.format("重试 [%d/%d]: %s", attempt, maxRetries, operationName);
    if (context != null && !context.isEmpty()) {
      msg += " | 上下文: " + context;
    }
    log("WARN", msg);
  }

  /** 记录成功 */
  private void logSuccess(String operationName, int attempt, Map<String, String> context) {
    String msg = "成功: " + operationName;
    if (attempt > 0) {
      msg += String.format(" (经过 %d 次重试)", attempt);
    }
    if (context != null && !context.isEmpty()) {
      msg += " | 上下文: " + context;
    }
    log("INFO", msg);
  }

  /** 记录失败 */
  private void logFailure(
      String operationName, int attempt, Exception error, Map<String, String> context) {
    String msg = String.format("失败 [%d/%d]: %s", attempt, maxRetries + 1, operationName);
    msg += String.format(" | 错误: %s: %s", error.getClass().getSimpleName(), error.getMessage());
    if (context != null && !context.isEmpty()) {
      msg += " | 上下文: " + context;
    }
    log("ERROR", msg);
  }

  /** 记录最终失败 */
  private void logFinalFailure(
      String operationName, int totalAttempts, Exception error, Map<String, String> context) {
    StringBuilder msg = new StringBuilder();
    msg.append("==============================\n");
    msg.append("最终失败: ").append(operationName).append("\n");
    msg.append("总尝试次数: ").append(totalAttempts).append("\n");
    msg.append("最后错误: ")
        .append(error.getClass().getSimpleName())
        .append(": ")
        .append(error.getMessage())
        .append("\n");
    if (context != null && !context.isEmpty()) {
      msg.append("上下文: ").append(context).append("\n");
    }
    msg.append("时间: ").append(LocalDateTime.now().format(dateFormatter)).append("\n");
    msg.append("==============================\n");

    log("CRITICAL", msg.toString());
  }

  /** 写入日志 */
  private void log(String level, String message) {
    String timestamp = LocalDateTime.now().format(dateFormatter);
    String logLine = String.format("[%s] [%s] %s", timestamp, level, message);

    // 输出到 stderr（避免破坏 GTP 协议）
    System.err.println(logLine);

    // 写入文件
    if (enableLogging) {
      try (FileWriter fw = new FileWriter(logFile, true);
          PrintWriter pw = new PrintWriter(fw)) {
        pw.println(logLine);
      } catch (IOException e) {
        System.err.println("无法写入日志文件: " + e.getMessage());
      }
    }
  }

  /** 测试方法 */
  public static void main(String[] args) {
    System.out.println("测试重试管理器...");

    RetryManager manager = new RetryManager(1, "test_retry.log", true);

    // 测试 1：成功的操作
    RetryResult<String> result1 = manager.executeWithRetry(() -> "成功结果", "测试成功操作");
    assert result1.success;
    assert "成功结果".equals(result1.result);
    System.out.println("✅ 测试 1 通过: " + result1.result);

    // 测试 2：第一次失败，第二次成功
    final int[] attemptCounter = {0};
    RetryResult<String> result2 =
        manager.executeWithRetry(
            () -> {
              attemptCounter[0]++;
              if (attemptCounter[0] == 1) {
                throw new RuntimeException("第一次尝试失败");
              }
              return "第二次成功";
            },
            "测试重试一次");
    assert result2.success;
    assert "第二次成功".equals(result2.result);
    System.out.println("✅ 测试 2 通过: " + result2.result);

    // 测试 3：总是失败
    Map<String, String> context = new HashMap<>();
    context.put("key", "value");
    RetryResult<String> result3 =
        manager.executeWithRetry(
            () -> {
              throw new RuntimeException("总是失败的操作");
            },
            "测试总是失败",
            context);
    assert !result3.success;
    assert result3.error instanceof RuntimeException;
    System.out.println("✅ 测试 3 通过: 正确处理了失败");

    System.out.println("\n所有测试通过！");
  }
}
